### DISCUSSION TITLE
Object Registration Issue

### DISCUSSION THREAD

**Discussion by jsevic** [idaholab/moose#25782] [Q&A Modules: General]
I have a custom kernel DiffusionCoupled that compiles and runs in the heat_conduction module. When I copy the .C and .h files to the phase_field directory and run make I am getting an error. I have updated the DiffusionCoupled kernel with
```
registerMooseObject("PhaseFieldApp", DiffusionCoupled);
```
to register my app.

When I run a simulation I get the following error thank you!

```
*** ERROR ***
A 'DiffusionCoupled' is not a registered object.

If you are trying to find this object in a dynamically loaded library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.

```

**Comment by GiudGiud** [comment]
Hello

heat_conduction module does not pull in phase field. You should run with the phase field executable instread

**Reply by jsevic** [reply]
At the moment I believe I am in the phase_field directory and executing phase_field-opt. Here is my current terminal. These two new kernels I made - DiffusionCoupled and JouleHeatingSource run in heat_conduction. Thx!


(moose) johnsevic@MBP-16 phase_field % make -j12
(moose) johnsevic@MBP-16 phase_field % ./phase_field-opt -i etpf-23_test_01.i

*** ERROR ***
A 'DiffusionCoupled' is not a registered object.

If you are trying to find this object in a dynamically loaded library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.

**Reply by GiudGiud** [reply]
I would think phase field pulls in heat_conduction. Weird. 
Can you `ldd <executable>` on linux? (or `otool -L` on macos)

you need to register these objects to the application you are using.

**Reply by jsevic** [reply]
I am running MacOS what file to run with the otool -L command? I tried phase_field-opt that did not seem to work?

Thx!

**Reply by GiudGiud** [reply]
`otool -L <executable>` should give us which libraries are linked to that executable.
And if the objects are is the heat conduction module and it s not linked in that s the problem

if not then it s got to be either a spelling issue on the app the objects are registered to
or you havent really recompiled, maybe the failure at compile-time is just a single line at the eend and it was missed. But that s unlikely

**Reply by jsevic** [reply]
This is what I get, it's not clear to me what should be looking for? I see tensor mechanics? Is there heat conduction? I have recently compiled. No errors are returned. Should I check the spelling of my custom kernels?
```
(moose) johnsevic@MBP-16 phase_field % otool -L phase_field-opt
phase_field-opt:
	/Users/johnsevic/projects/moose/modules/phase_field/test/lib/libphase_field_test-opt.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/Users/johnsevic/projects/moose/modules/phase_field/lib/libphase_field-opt.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/Users/johnsevic/projects/moose/modules/tensor_mechanics/lib/libtensor_mechanics-opt.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/Users/johnsevic/projects/moose/framework/libmoose-opt.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/Users/johnsevic/projects/moose/framework/contrib/pcre/libpcre-opt.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/Users/johnsevic/projects/moose/framework/contrib/hit/libhit-opt.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libwasplsp.03.dylib (compatibility version 3.0.0, current version 3.1.3)
	@rpath/libwasphalite.03.dylib (compatibility version 3.0.0, current version 3.1.3)
	@rpath/libwaspson.03.dylib (compatibility version 3.0.0, current version 3.1.3)
	@rpath/libwaspcore.03.dylib (compatibility version 3.0.0, current version 3.1.3)
	@rpath/libwaspexpr.03.dylib (compatibility version 3.0.0, current version 3.1.3)
	@rpath/libwaspjson.03.dylib (compatibility version 3.0.0, current version 3.1.3)
	@rpath/libwaspddi.03.dylib (compatibility version 3.0.0, current version 3.1.3)
	@rpath/libwasphive.03.dylib (compatibility version 3.0.0, current version 3.1.3)
	@rpath/libwasphit.03.dylib (compatibility version 3.0.0, current version 3.1.3)
	@rpath/libwaspsiren.03.dylib (compatibility version 3.0.0, current version 3.1.3)
	@rpath/../libmesh/lib/libmesh_opt.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/../libmesh/lib/libtimpi_opt.11.dylib (compatibility version 12.0.0, current version 12.0.0)
	@rpath/libhdf5_cpp.200.dylib (compatibility version 203.0.0, current version 203.0.0)
	@rpath/libvtkIOCore-9.2.1.dylib (compatibility version 1.0.0, current version 9.2.6)
	@rpath/libvtkCommonCore-9.2.1.dylib (compatibility version 1.0.0, current version 9.2.6)
	@rpath/libvtkCommonDataModel-9.2.1.dylib (compatibility version 1.0.0, current version 9.2.6)
	@rpath/libvtkFiltersCore-9.2.1.dylib (compatibility version 1.0.0, current version 9.2.6)
	@rpath/libvtkIOXML-9.2.1.dylib (compatibility version 1.0.0, current version 9.2.6)
	@rpath/libvtkImagingCore-9.2.1.dylib (compatibility version 1.0.0, current version 9.2.6)
	@rpath/libvtkIOImage-9.2.1.dylib (compatibility version 1.0.0, current version 9.2.6)
	@rpath/libvtkImagingMath-9.2.1.dylib (compatibility version 1.0.0, current version 9.2.6)
	@rpath/libvtkIOParallelXML-9.2.1.dylib (compatibility version 1.0.0, current version 9.2.6)
	@rpath/libvtkParallelMPI-9.2.1.dylib (compatibility version 1.0.0, current version 9.2.6)
	@rpath/libvtkParallelCore-9.2.1.dylib (compatibility version 1.0.0, current version 9.2.6)
	@rpath/libvtkCommonExecutionModel-9.2.1.dylib (compatibility version 1.0.0, current version 9.2.6)
	@rpath/libvtksys-9.2.1.dylib (compatibility version 1.0.0, current version 9.2.6)
	@rpath/libz.1.dylib (compatibility version 1.0.0, current version 1.2.13)
	@rpath/libslepc.3.019.dylib (compatibility version 3.19.0, current version 3.19.1)
	@rpath/libpetsc.3.019.dylib (compatibility version 3.19.0, current version 3.19.3)
	@rpath/libHYPRE-2.29.0.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.100.3)
	@rpath/libstrumpack.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libscalapack.2.2.dylib (compatibility version 2.2.0, current version 2.2.1)
	@rpath/libsuperlu_dist.8.dylib (compatibility version 8.0.0, current version 8.1.2)
	@rpath/libparmetis.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libmetis.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libhdf5_hl.200.dylib (compatibility version 202.0.0, current version 202.0.0)
	@rpath/libhdf5.200.dylib (compatibility version 203.0.0, current version 203.0.0)
	@rpath/libmpifort.12.dylib (compatibility version 15.0.0, current version 15.2.0)
	@rpath/libpmpi.12.dylib (compatibility version 15.0.0, current version 15.2.0)
	@rpath/libgfortran.5.dylib (compatibility version 6.0.0, current version 6.0.0)
	@rpath/libquadmath.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libc++.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libmpi.12.dylib (compatibility version 15.0.0, current version 15.2.0)
	@rpath/libmpicxx.12.dylib (compatibility version 15.0.0, current version 15.2.0)
	@rpath/libomp.dylib (compatibility version 5.0.0, current version 5.0.0)
```

**Reply by GiudGiud** [reply]
but exactly. There is not heat conduction

so anything you compiled in that module is not being pulled in

you can either modify modules/modules.mk to make phase_field pull heat_conduction or you can create an app that pulls both modules

or move your code from HC to PF

**Reply by jsevic** [reply]
OK so two questions, at the moment in this instance I had JouleHeatingSource in my .i file, however in another instance I had my two new kernels, DiffusionCoupled and JouleSourceCoupled, and in the heat_conduction module these worked fine. When I load these into the src and include directories of phase_field, and change the registration command to PhaseFieldApp, I still get the run time error of function not found? So, essentially, I have moved my code from HC to PF. I did double-check spelling. I will try editing the modules.mk to see what happens. What about moving all the PF to HC? Thx!

**Reply by GiudGiud** [reply]
Should have worked. Maybe the registration system is being stubborn. Did you git commit then git clean and try again?

**Comment by jsevic** [comment]
Ok let me try that. I am also going to triple check my spelling.

One more question, at the moment I have been running everything from the MOOSE install directory, I had some issues making a new app. Maybe I should try making a new app also with HC and PF?

**Reply by jsevic** [reply]
Well this is interesting I tried commit and clean but no effect. Then, I thought I would try creating a new app, even though previously this was causing some grief.

I made a new app called xenon and set phase_field and heat_conduction to yes in the makefile. This time I see heat_conduction and phase_field show up when I run the otool command as you suggested earlier.

(moose) johnsevic@MBP-16 xenon % otool -L xenon-opt

I have also added my custom kernels and register them to xenonApp. However, when I ran a PF test case I got the following result, suggesting that phase_field was not linked. Then I was poking around on the MOOSE web-site and I found this command and everything works as I was hoping for, I have my own kernels, HC, and PF all in one app. I am not sure overall what happened but it looks ok now.


[Problem]
    register_objects_from = 'PhaseFieldApp'
    library_path = '/Users/johnsevic/projects/moose/modules/phase_field/lib'
[]

*** ERROR ***
A 'SplitCHWRes' is not a registered object.

If you are trying to find this object in a dynamically loaded library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.

**Reply by GiudGiud** [reply]
```
[Problem]
register_objects_from = 'PhaseFieldApp'
library_path = '/Users/johnsevic/projects/moose/modules/phase_field/lib'
[]
```

this is registering dynamically the objects from this app.
It should not be required. It s a valid way of brining objects from a module you did NOT build with. It should not be necessary. 

Can you attach the output of your executable --json ?
This will say which app each object is registered too. And I ll be able to understand this more, maybe

**Reply by jsevic** [reply]
I ran xenon-opt --json and the terminal produces hundreds of lines of text is that what you want?

Also I did just note that I have been using the TimeDerivative operator but adopted the HeatConductionTimeDerivative to add rho and cp and then I got an error that said HeatConductionTimeDerivative was not linked to my code. Thx!

**Reply by GiudGiud** [reply]
Yes. You ll need to dump it to file using 
&> file_name 
After the command

**Reply by jsevic** [reply]
OK thanks here you go...
[etpf-23_test_01.txt](https://github.com/idaholab/moose/files/13166762/etpf-23_test_01.txt)

**Reply by GiudGiud** [reply]
all 1200 objects here are registered to MooseApp. No phase field no heat conduction

can you please paste your Makefile

**Reply by jsevic** [reply]
Here you go, and do you need the make file from the xenon build too? That is where I set PF and HC to "yes"
```
###############################################################################
################### MOOSE Application Standard Makefile #######################
###############################################################################
#
# Optional Environment variables
# MOOSE_DIR     - Root directory of the MOOSE project
# FRAMEWORK_DIR - Location of the MOOSE framework
#
###############################################################################
EXAMPLE_DIR        ?= $(shell dirname `pwd`)
MOOSE_DIR          ?= $(shell dirname $(EXAMPLE_DIR))
FRAMEWORK_DIR      ?= $(MOOSE_DIR)/framework
###############################################################################

TEST := test_ignore

# framework
include $(FRAMEWORK_DIR)/build.mk
include $(FRAMEWORK_DIR)/moose.mk

APPLICATION_NAME := ex03

# dep apps
APPLICATION_DIR    := $(shell pwd)
APPLICATION_NAME   := phase_field
BUILD_EXEC         := yes
DEP_APPS           := $(shell $(FRAMEWORK_DIR)/scripts/find_dep_apps.py $(APPLICATION_NAME))
include            $(FRAMEWORK_DIR)/app.mk

# Include dependency files for this example
ex_srcfiles := $(shell find $(APPLICATION_DIR) -name "*.C")
ex_deps     := $(patsubst %.C, %.$(obj-suffix).d, $(ex_srcfiles))
-include $(ex_deps)

###############################################################################
# Additional special case targets should be added here

test: all 
	$(call TEST_exodiff,ex03.i,out.e)
```

**Reply by GiudGiud** [reply]
which Makefile is that?

yes XenonApp as well

**Reply by jsevic** [reply]
Here is the makefile I used to make my new app xenon, and then I have turned on PF and HC?

**Reply by GiudGiud** [reply]
nothing wrong here or the one below

and you mentioned otool shows heat conduction

**Comment by jsevic** [comment]
```
###############################################################################
################### MOOSE Application Standard Makefile #######################
###############################################################################
#
# Optional Environment variables
# MOOSE_DIR        - Root directory of the MOOSE project
#
###############################################################################
# Use the MOOSE submodule if it exists and MOOSE_DIR is not set
MOOSE_SUBMODULE    := $(CURDIR)/moose
ifneq ($(wildcard $(MOOSE_SUBMODULE)/framework/Makefile),)
  MOOSE_DIR        ?= $(MOOSE_SUBMODULE)
else
  MOOSE_DIR        ?= $(shell dirname `pwd`)/moose
endif

# framework
FRAMEWORK_DIR      := $(MOOSE_DIR)/framework
include $(FRAMEWORK_DIR)/build.mk
include $(FRAMEWORK_DIR)/moose.mk

################################## MODULES ####################################
# To use certain physics included with MOOSE, set variables below to
# yes as needed.  Or set ALL_MODULES to yes to turn on everything (overrides
# other set variables).

ALL_MODULES                 := no

CHEMICAL_REACTIONS          := no
CONTACT                     := no
ELECTROMAGNETICS            := no
EXTERNAL_PETSC_SOLVER       := no
FLUID_PROPERTIES            := no
FSI                         := no
FUNCTIONAL_EXPANSION_TOOLS  := no
GEOCHEMISTRY                := no
HEAT_CONDUCTION             := yes
LEVEL_SET                   := no
MISC                        := no
NAVIER_STOKES               := no
OPTIMIZATION                := no
PERIDYNAMICS                := no
PHASE_FIELD                 := yes
POROUS_FLOW                 := no
RAY_TRACING                 := no
REACTOR                     := no
RDG                         := no
RICHARDS                    := no
STOCHASTIC_TOOLS            := no
THERMAL_HYDRAULICS          := no
TENSOR_MECHANICS            := no
XFEM                        := no

include $(MOOSE_DIR)/modules/modules.mk
###############################################################################

# dep apps
APPLICATION_DIR    := $(CURDIR)
APPLICATION_NAME   := xenon
BUILD_EXEC         := yes
GEN_REVISION       := no
include            $(FRAMEWORK_DIR)/app.mk

###############################################################################
# Additional special case targets should be added here
```

**Reply by jsevic** [reply]
Should I try doing a new fresh MOOSE install? Any other workarounds?

**Reply by GiudGiud** [reply]
updating is not a bad idea in any case.
make sure to save all your work in both repos  first and clean them with `git clean -xfd` (will remove everything)

**Comment by jsevic** [comment]
For the moment here is what I did, since I needed the HeatConductionTimeDerivative operator HC I made my own heat conduction operator coupled to material properties and now everything for electrothermal phase field seems to be working. I will also try updating MOOSE. Thx!
